### DISCUSSION TITLE
How to build a query for a specific model?

### DISCUSSION THREAD

**Discussion by AlejandroAkbal** [rappasoft/laravel-livewire-tables#392] [Q&A]
I'm trying to show a user its projects, instead of every single project that there exists.

I have three models, "User", "Project" and "Link".

A "User" has many "Project"

"Project" has many "Link"

Now, I've tried the following LiveWire component:

```php
<?php

namespace App\Http\Livewire;

use App\Models\Link;
use App\Models\Project;
use Illuminate\Database\Eloquent\Builder;
use Rappasoft\LaravelLivewireTables\DataTableComponent;
use Rappasoft\LaravelLivewireTables\Views\Column;

class LinksDataTable extends DataTableComponent
{
    // Data
    public Project $parentProject;

    public function columns(): array
    {
        return [
            Column::make('URL Slug', 'url_slug')
                ->sortable()
                ->searchable(),

            Column::make('Original URL', 'original_url')
                ->sortable()
                ->searchable(),
        ];
    }

    public function query(): Builder
    {
        return Link::where('project_id', $this->parentProject->id);
    }

    public function mount($parentProject)
    {
        // Assign project
        $this->parentProject = $parentProject;
    }
}
```

The code above works, but when using the search form, it dies with the following SQL Exception:

```sql
SQLSTATE[42S22]: Column not found: 1054 Unknown column 'links.status' in 'where clause' (SQL: select count(*) as aggregate from `links` where `project_id` = 1 and (`links`.`url_slug` like %atq% or `links`.`original_url` like %atq% or `links`.`status` like %atq%)) 
```

The question is, what am I doing wrong?

**Comment by rappasoft** [comment]
Where does `status` come into play in this component? I don't see it.

**Reply by AlejandroAkbal** [reply]
Sorry, I removed some code since I thought it wasn't doing anything.

I thought columns without the "column" parameter were just visual and did nothing.

Full code:

```php
<?php

namespace App\Http\Livewire;

use App\Models\Link;
use App\Models\Project;
use Illuminate\Database\Eloquent\Builder;
use Rappasoft\LaravelLivewireTables\DataTableComponent;
use Rappasoft\LaravelLivewireTables\Views\Column;

class LinksDataTable extends DataTableComponent
{
    // Data
    public Project $parentProject;

    // Options
    public ?int $searchFilterDebounce = 350;

    public function columns(): array
    {
        return [
            Column::make('URL Slug', 'url_slug')
                ->sortable()
                ->searchable(),

            Column::make('Original URL', 'original_url')
                ->sortable()
                ->searchable(),

            Column::make('Status')
                ->sortable()
                ->searchable(),

            Column::make('Analytics')
                ->sortable(),

            Column::make('Actions')
        ];
    }

    public function query(): Builder
    {
        return Link::where('project_id', $this->parentProject->id);
    }

    public function mount($parentProject)
    {
        // Assign project
        $this->parentProject = $parentProject;
    }
}
```

Removing those columns solve my issue. 

Sorry about that, I was too focused on the "query" function that I didn't actually associate the SQL Exception to the column.

**Reply by rappasoft** [reply]
Is there a status column on that model? Because you have it as searchable, so the query is going to include it.

You also have it as sortable, so if you click the table header it will do an orderBy on that column and will probably crash as well if the column doesn't exist.

**Reply by AlejandroAkbal** [reply]
No, that is the problem.

I wanted to make that row display a badge.

Similar to the "active" row on these screenshots from you.
https://github.com/rappasoft/laravel-livewire-tables/discussions/232#discussioncomment-650175

Could you tell me how to do that?

I assume they are custom views, created this way, right?
https://github.com/rappasoft/laravel-livewire-tables/wiki/Using-the-built-in-cell-formatting#you-could-also-return-a-view

**Reply by rappasoft** [reply]
Well, that column in the demo works off of an 'active' column on the model. Otherwise, how would it know?

https://gist.github.com/rappasoft/948adf542307b8f620d53c7c7e735d3c#file-laravel-livewire-tables-demo-table-php-L35

It's just a simple view partial that checks true or false and displays an Check or X.

I would assume if you want a status column, then you have it on either that model or a relationships model because the query needs to know what to do when you search or sort, and how to display it based on its value.

**Reply by AlejandroAkbal** [reply]
Yeah, since I'm in the early stages of building it, I didn't quite think of how to do it.
I will probably make it a model method.

Anyway, thank you very much for all your help, Anthony. 🙏

You're wonderful.

**Reply by rappasoft** [reply]
Thanks! I would just make it a column on the model:

If it has more than 2 values, maybe an enum: 'active, pending, complete', etc.

If it has either on or off, such as the 'active' case here, then just make it a boolean.
